File: python/canvastekk_workflow_sdk/workflow/resolver.py
"""
Input Resolver

Resolves a node's inputs from static params + incoming edge outputs.
Supports dot-notation traversal for nested output extraction.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from canvastekk_workflow_sdk.workflow.models import WorkflowDefinitionSpec


def resolve_inputs(
    node_id: str,
    spec: WorkflowDefinitionSpec,
    node_outputs: dict[str, dict[str, Any]],
) -> dict[str, Any]:
    """Resolves a node's inputs from static params + incoming edge outputs.

    Merges static inputs with values from connected upstream nodes.
    Supports dot-notation traversal for nested output extraction.

    Args:
        node_id: Target node ID
        spec: Workflow definition spec
        node_outputs: Map of node ID -> outputs from prior executions

    Returns:
        Dict of resolved input values for the target node

    Raises:
        KeyError: If node not found in spec or output key resolution fails
    """
    try:
        node = next(n for n in spec.nodes if n.id == node_id)
    except StopIteration:
        raise KeyError(f"Node '{node_id}' not found in spec") from None
    resolved: dict[str, Any] = dict(node.inputs)

    incoming = [e for e in spec.edges if e.to_node == node_id]
    for edge in incoming:
        source_outputs = node_outputs.get(edge.from_node, {})
        value = _resolve_output(source_outputs, edge.from_output, edge.from_node)
        if edge.to_input:
            resolved[edge.to_input] = value
        elif isinstance(value, dict):
            resolved.update(value)

    return resolved


def _resolve_output(
    source_outputs: dict[str, Any],
    from_output: str,
    from_node: str = "",
) -> Any:
    """Resolves a single output value from source node outputs.

    Supports dot-notation traversal for nested objects.

    Args:
        source_outputs: Source node's output object
        from_output: Output key path (empty string returns all outputs)
        from_node: Optional source node ID for error messages

    Returns:
        Resolved output value

    Raises:
        KeyError: If output key not found or path is invalid
    """
    if not from_output:
        return source_outputs

    if from_output in source_outputs:
        return source_outputs[from_output]
    if "." in from_output:
        return _walk_dot_path(source_outputs, from_output, from_node)
    node_ctx = f" from node '{from_node}'" if from_node else ""
    raise KeyError(
        f"Cannot resolve from_output '{from_output}'{node_ctx}; "
        f"available keys: {sorted(source_outputs.keys())}"
    )


def _walk_dot_path(data: dict[str, Any], path: str, from_node: str = "") -> Any:
    """Walks a dot-notation path through a nested object.

    Args:
        data: Object to traverse
        path: Dot-separated path (e.g., "output.nested.value")
        from_node: Optional source node ID for error messages

    Returns:
        Value at the path endpoint

    Raises:
        KeyError: If path is invalid, hits a non-object, or key not found
    """
    current: Any = data
    for segment in path.split("."):
        if not segment:
            raise KeyError(f"Invalid dot-path '{path}' (empty segment)")
        if not isinstance(current, dict):
            raise KeyError(
                f"Cannot walk dot-path '{path}': segment '{segment}' hits non-dict"
            )
        if segment not in current:
            node_ctx = f" from node '{from_node}'" if from_node else ""
            raise KeyError(f"Dot-path '{path}': segment '{segment}' not found{node_ctx}")
        current = current[segment]
    return current

File: typescript/src/workflow/resolver.ts
import type { WorkflowDefinitionSpec } from "./models.js";

/**
 * Error thrown when workflow input/output resolution fails.
 *
 * Includes an error code and optional node ID for debugging.
 */
export class ResolverError extends Error {
  /** Error code categorizing the failure type. */
  readonly code: "NODE_NOT_FOUND" | "KEY_NOT_FOUND" | "INVALID_PATH" | "TRAVERSAL_ERROR";
  /** Optional node ID where the error occurred (for debugging). */
  readonly nodeId?: string;

  /**
   * Creates a new ResolverError.
   *
   * @param message - Human-readable error message
   * @param code - Error type code
   * @param nodeId - Optional node ID where the error occurred
   */
  constructor(
    message: string,
    code: "NODE_NOT_FOUND" | "KEY_NOT_FOUND" | "INVALID_PATH" | "TRAVERSAL_ERROR",
    nodeId?: string,
  ) {
    super(message);
    this.name = "ResolverError";
    this.code = code;
    this.nodeId = nodeId;
  }
}

/**
 * Type guard to check if a value is a plain object (Record).
 *
 * @param value - Value to check
 * @returns True if the value is a non-null object and not an array
 */
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Resolves a node's inputs from static params + incoming edge outputs.
 *
 * @param nodeId - Target node ID
 * @param spec - Workflow definition spec
 * @param nodeOutputs - Map of node ID → outputs from prior executions
 * @returns Resolved inputs for the target node
 * @throws {ResolverError} When node not found or output key resolution fails
 */
export function resolveInputs(
  nodeId: string,
  spec: WorkflowDefinitionSpec,
  nodeOutputs: Record<string, Record<string, unknown>>,
): Record<string, unknown> {
  const node = spec.nodes.find((n) => n.id === nodeId);
  if (!node) throw new ResolverError(`Node not found: ${nodeId}`, "NODE_NOT_FOUND", nodeId);

  const resolved: Record<string, unknown> = { ...node.inputs };

  const incoming = spec.edges.filter((e) => e.to_node === nodeId);
  for (const edge of incoming) {
    const sourceOutputs = nodeOutputs[edge.from_node] ?? {};
    const value = resolveOutput(sourceOutputs, edge.from_output, edge.from_node);
    if (edge.to_input) {
      resolved[edge.to_input] = value;
    } else if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      Object.assign(resolved, value);
    }
  }

  return resolved;
}

/**
 * Resolves a single output value from source node outputs.
 *
 * Supports dot-notation traversal for nested objects.
 *
 * @param sourceOutputs - Source node's output object
 * @param fromOutput - Output key path (empty string returns all outputs)
 * @param fromNode - Optional source node ID for error messages
 * @returns Resolved output value
 * @throws {ResolverError} When output key not found or path is invalid
 */
function resolveOutput(
  sourceOutputs: Record<string, unknown>,
  fromOutput: string,
  fromNode?: string,
): unknown {
  if (!fromOutput) return sourceOutputs;

  if (fromOutput in sourceOutputs) return sourceOutputs[fromOutput];

  if (fromOutput.includes(".")) return walkDotPath(sourceOutputs, fromOutput, fromNode);

  const nodeCtx = fromNode ? ` from node '${fromNode}'` : "";
  throw new ResolverError(
    `Cannot resolve from_output '${fromOutput}'${nodeCtx}; available keys: ${sortedKeys(sourceOutputs)}`,
    "KEY_NOT_FOUND",
    fromNode,
  );
}

/**
 * Walks a dot-notation path through a nested object.
 *
 * @param data - Object to traverse
 * @param path - Dot-separated path (e.g., "output.nested.value")
 * @param fromNode - Optional source node ID for error messages
 * @returns Value at the path endpoint
 * @throws {ResolverError} When path is invalid, hits a non-object, or key not found
 */
function walkDotPath(data: Record<string, unknown>, path: string, fromNode?: string): unknown {
  let current: unknown = data;
  for (const segment of path.split(".")) {
    if (!segment) throw new ResolverError(`Invalid dot-path '${path}' (empty segment)`, "INVALID_PATH", fromNode);
    if (!isRecord(current)) {
      throw new ResolverError(
        `Cannot walk dot-path '${path}': segment '${segment}' hits non-dict`,
        "TRAVERSAL_ERROR",
        fromNode,
      );
    }
    if (!(segment in current)) {
      const nodeCtx = fromNode ? ` from node '${fromNode}'` : "";
      throw new ResolverError(
        `Dot-path '${path}': segment '${segment}' not found${nodeCtx}`,
        "KEY_NOT_FOUND",
        fromNode,
      );
    }
    current = current[segment];
  }
  return current;
}

/**
 * Returns a sorted list of an object's keys as a string for error messages.
 *
 * @param obj - Object to extract keys from
 * @returns Sorted keys as a bracketed string
 */
function sortedKeys(obj: Record<string, unknown>): string {
  return `[${Object.keys(obj).sort().join(", ")}]`;
}
